tests/extensibleLM/ExtensibleLoadManagerTest.cc

```diff
@@ -0,0 +1,210 @@
+
+    Semaphore firstUnloadSemaphore(0);
+    Semaphore secondUnloadSemaphore(0);
+    Semaphore halfPubWaitSemaphore(0);
+    const int msgCount = 10;
+    int produced = 0;
+    auto produce = [&producer, &produced, &firstUnloadSemaphore, &secondUnloadSemaphore,
+                    &halfPubWaitSemaphore]() {
+        int i = 0;
+        while (i < msgCount && checkTime()) {
+            if (i == 3) {
+                firstUnloadSemaphore.acquire();
+            }
+
+            if (i == 5) {
+                halfPubWaitSemaphore.release();
+            }
+
+            if (i == 8) {
+                secondUnloadSemaphore.acquire();
+            }
+
+            std::string content = std::to_string(i);
+            const auto msg = MessageBuilder().setContent(content).build();
+            while (checkTime()) {
+                Result sendResult = producer.send(msg);
+                if (sendResult != ResultOk) {
+                    continue;
+                }
+                break;
+            }
+            std::cout << "produced index:" << i << std::endl;
+            produced++;
+            i++;
+        }
+        std::cout << "producer finished" << std::endl;
+    };
+
+    int consumed = 0;
+    auto consume = [&consumer, &consumed]() {
+        Message receivedMsg;
+        int i = 0;
+        while (i < msgCount && checkTime()) {
+            while (checkTime()) {
+                Result receiveResult =
+                    consumer.receive(receivedMsg, 1000);  // Assumed that we wait 1000 ms for each message
+                if (receiveResult != ResultOk) {
+                    continue;
+                }
+                std::cout << "received index:" << i << std::endl;
+                break;
+            }
+            int id = std::stoi(receivedMsg.getDataAsString());
+            if (id < i) {
+                continue;
+            }
+            while (checkTime()) {
+                Result ackResult = consumer.acknowledge(receivedMsg);
+                if (ackResult != ResultOk) {
+                    continue;
+                }
+                std::cout << "acked index:" << i << std::endl;
+                break;
+            }
+            consumed++;
+            i++;
+        }
+        std::cout << "consumer finished" << std::endl;
+    };
+
+    std::thread produceThread(produce);
+    std::thread consumeThread(consume);
+
+    auto unload = [&client, &consumer, &producer] {
+        auto clientImplPtr = PulsarFriend::getClientImplPtr(client);
+        auto &consumerImpl = PulsarFriend::getConsumerImpl(consumer);
+        auto &producerImpl = PulsarFriend::getProducerImpl(producer);
+        uint64_t lookupCountBeforeUnload;
+        std::string destinationBroker;
+        while (checkTime()) {
+            // make sure producers and consumers are ready
+            while (checkTime() && !consumerImpl.isConnected() && !producerImpl.isConnected()) {
+                sleep(1);
+            }
```

Comment:
Replace the explicit `sleep` call with the `waitUntil` function from `tests/WaitUtils.h`. For example, set 5s as the timeout:

```c++
            ASSERT_TRUE(waitUntil(std::chrono::seconds(5), [&] {
                return checkTime() && !consumerImpl.isConnected() && !producerImpl.isConnected();
            }));
```

Reply:
Sure. Thanks for the suggestion.